libretro-common/include/lists/linked_list.h

```diff
@@ -0,0 +1,298 @@
+ * "list" is NULL.
+ *
+ * @param list list to add the element to
+ * @param value new value to add to the linked list
+ */
+void linked_list_add(linked_list_t *list, void *value);
+
+/**
+ * @brief inserts a value into the linked list
+ *
+ * Inserts a value into the linked list at the specified index. Does
+ * nothing if "list" is NULL.
+ *
+ * @param list list to insert the value into
+ * @param index index where the value should be inserted at (can be equal to list size)
+ * @param value value to insert into the linked list
+ */
+void linked_list_insert(linked_list_t *list, size_t index, void *value);
+
+/**
+ * @brief Get the value in the linked list at the provided index.
+ *
+ * Return the value vstored in the linked list at the provided index. Does
+ * nothing if "list" is NULL.
+ *
+ * @param list list to get the value from
+ * @param index index of the value to return
+ * @return value in the list at the provided index
+ */
+void *linked_list_get(linked_list_t *list, size_t index);
+
+/**
+ * @brief Get the first value that is matched by the provided function
+ *
+ * Return the first value that the function matches. The matches function
+ * parameters are value from the linked list and the provided state.
+ *
+ * @param list list to get the value from
+ * @param matches function to test the values with
+ * @param state state to pass to the matches function
+ * @return first value that matches otherwise NULL
+ */
+void *linked_list_get_first_matching(linked_list_t *list, bool (*matches)(void *item, void *state), void *state);
+
+/**
+ * @brief Get the last value that is matched by the provided function
+ *
+ * Return the last value that the function matches. The matches function
+ * parameters are value from the linked list and the provided state.
+ *
+ * @param list list to get the value from
+ * @param matches function to test the values with
+ * @param state state to pass to the matches function
+ * @return last value that matches otherwise NULL
+ */
+void *linked_list_get_last_matching(linked_list_t *list, bool (*matches)(void *item, void *state), void *state);
+
+/**
+ * @brief Remove the element at the provided index
+ *
+ * Removes the element of the linked list at the provided index.
+ *
+ * @param list linked list to remove the element from
+ * @param index index of the element to remove
+ * @return value of the element that was removed, NULL if list is NULL or
+ *         index is invalid
+ */
+void *linked_list_remove_at(linked_list_t *list, size_t index);
+
+/**
+ * @brief Remove the first element with the provided value
+ *
+ * Removes the first element with a value equal to the provided value.
+ * Does nothing if "list" is NULL.
+ *
+ * @param list linked list to remove the element from
+ * @param value value of the element to remove
+ * @return value if a matching element was removed
+ */
+void *linked_list_remove_first(linked_list_t *list, void *value);
```

Comment:
Nit: call them pop_first and pop_last maybe? Since they do not really remove but rather return the element *and* remove it from the list.

Reply:
I'm not sure about this. Your suggestion doesn't fit for all of the similar functions (such as `linked_list_remove_all()`). I also see `pop` as an operation on the first or last elements, not really elements in the middle.

I see three choices:
1. Rename the functions except for `linked_list_remove_all` and change it to not return anything. Simlar thing with the "matching" functions. To be fair, only `linked_list_remove_first_matching()` and `linked_list_remove_last_matching()` really need to return the removed element.
2. Leave the names alone
3. Stop returning the removed value. This makes `linked_list_remove_first_matching()` less useful though.